flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/typeutils/BaseMapSerializer.java

```diff
@@ -0,0 +1,273 @@
+ * the License.	You may obtain a copy of the License at
+ *
+ *		http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.typeutils;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializerSchemaCompatibility;
+import org.apache.flink.api.common.typeutils.TypeSerializerSnapshot;
+import org.apache.flink.api.java.typeutils.runtime.DataInputViewStream;
+import org.apache.flink.api.java.typeutils.runtime.DataOutputViewStream;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.table.dataformat.BaseMap;
+import org.apache.flink.table.dataformat.BinaryArray;
+import org.apache.flink.table.dataformat.BinaryArrayWriter;
+import org.apache.flink.table.dataformat.BinaryMap;
+import org.apache.flink.table.dataformat.BinaryWriter;
+import org.apache.flink.table.dataformat.GenericMap;
+import org.apache.flink.table.types.InternalSerializers;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.util.SegmentsUtil;
+import org.apache.flink.util.InstantiationUtil;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Serializer for {@link BaseMap}.
+ */
+public class BaseMapSerializer extends TypeSerializer<BaseMap> {
+
+	private final LogicalType keyType;
+	private final LogicalType valueType;
+
+	private final TypeSerializer keySerializer;
+	private final TypeSerializer valueSerializer;
+
+	private transient BinaryArray reuseKeyArray;
+	private transient BinaryArray reuseValueArray;
+	private transient BinaryArrayWriter reuseKeyWriter;
+	private transient BinaryArrayWriter reuseValueWriter;
+
+	public BaseMapSerializer(LogicalType keyType, LogicalType valueType) {
+		this.keyType = keyType;
+		this.valueType = valueType;
+
+		this.keySerializer = InternalSerializers.create(keyType, new ExecutionConfig());
+		this.valueSerializer = InternalSerializers.create(valueType, new ExecutionConfig());
+	}
+
+	@Override
+	public boolean isImmutableType() {
+		return false;
+	}
+
+	@Override
+	public TypeSerializer<BaseMap> duplicate() {
+		return new BaseMapSerializer(keyType, valueType);
+	}
+
+	@Override
+	public BaseMap createInstance() {
+		return new BinaryMap();
+	}
+
+	@Override
+	public BaseMap copy(BaseMap from) {
+		if (from instanceof GenericMap) {
+			Map<Object, Object> fromMap = ((GenericMap) from).getMap();
+			HashMap<Object, Object> toMap = new HashMap<>();
```

Comment:
Here we are using HashMap, so there are some constraints on the key type (the hashCode & equals methods must be implemented properly). 

Are such constraints acceptable, or we should use a list of key-value pairs internally?

Reply:
No,
First: only `IdentityConverter` will use `GenericMap`, others will use `BinaryMap`.
Second: Just wrap user `Map`, users need to guarantee their logic.

Reply:
"Just wrap user Map, users need to guarantee their logic."
-> the users just guarantee that their Map works correctly according to the Map interface, but it does not mean their Map would work correctly when wrapped in a HashMap. Right?

When that is the case, it may cause some weird problem which is hard to locate and debug. So this is not a reliable solution. 

At the very least, we should write some comment in the JavaDoc about this explicitly.

Reply:
This has nothing to do with `GenericMap`. If the user gives a wrong `Map`, it's wrong to turn it into a `BinaryMap` too. It is totally wrong...

Reply:
The user can provide a "correct" Map, but when the key value is wrapped in a HashMap, it can be a "wrong" Map.

Let me illustrate with an example. Suppose the key type does not implement "hashCode" and "equals" functions correctly (suppose for simplicity, hashCode always returns 0, and equals always returns true).

The input GenericMap can be based on a TreeMap, with a proper comparator to test key equality. 
The input map is a "correct" map, because it implements get/put methods correctly. 

However, when we insert the key/value pairs of the TreeMap into a HashMap, problems will occur. For this example, only one key/value pair will finally exist in the HashMap, which is an unexpected behavior.

Reply:
You are right,
`DataFormatConverter` is not wrong, but `copy` maybe have some trouble.
In fact, Runtime `MapSerializer` also has this problem.
I will add some comment.